lib/useOnyx.ts

```diff
@@ -0,0 +1,208 @@
+import {deepEqual} from 'fast-equals';
+import {useCallback, useEffect, useRef, useSyncExternalStore} from 'react';
+import Onyx from './Onyx';
+import type {CollectionKeyBase, KeyValueMapping, OnyxCollection, OnyxEntry, OnyxKey, Selector} from './types';
+import usePrevious from './usePrevious';
+
+type OnyxValue<TKey extends OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
+
+type UseOnyxOptions<TKey extends OnyxKey, TReturnData> = {
+    /**
+     * Determines if this key in this subscription is safe to be evicted.
+     */
+    canEvict?: boolean;
+
+    /**
+     * If set to false, then no data will be prefilled into the component.
+     */
+    initWithStoredValues?: boolean;
+
+    /**
+     * TODO: Check if we still need this flag and associated logic.
+     */
+    // allowStaleData?: boolean;
+
+    /**
+     * Sets an initial value to be returned by the hook during the first render.
+     */
+    initialValue?: OnyxValue<TKey>;
+
+    /**
+     * If included, this will be used to subscribe to a subset of an Onyx key's data.
+     * Using this setting on `useOnyx` can have very positive performance benefits because the component will only re-render
+     * when the subset of data changes. Otherwise, any change of data on any property would normally
+     * cause the component to re-render (and that can be expensive from a performance standpoint).
+     */
+    selector?: Selector<TKey, unknown, TReturnData>;
+};
+
+type FetchStatus = 'loading' | 'loaded';
+
+type UseOnyxDataValue<TKey extends OnyxKey, TValue> = TValue extends OnyxValue<TKey> ? TValue : TKey extends CollectionKeyBase ? NonNullable<OnyxCollection<TValue>> : TValue;
+
+type UseOnyxData<TKey extends OnyxKey, TValue> = {
+    value: UseOnyxDataValue<TKey, TValue>;
+    status: FetchStatus;
+};
+
+function isCollectionMemberKey<TKey extends OnyxKey>(key: TKey): boolean {
```

Comment:
Can we reuse `isCollectionMemberKey` from Onyx.js here?

Reply:
I'm not sure, I saw that we already have a `isCollectionMemberKey` in `Onyx.js` but it actually expects two params (`collectionKey` and `key`), while I have only `key` in the hook. I will check again if we can reuse it somehow

Reply:
Actually I think we can reuse it, will do it